README.md
<SPAN ALIGN="CENTER">

[![homebridge-myq2: Native HomeKit support for myQ garage door openers and other devices](https://raw.githubusercontent.com/hjdhjd/homebridge-myq2/master/homebridge-myq.svg)](https://github.com/hjdhjd/homebridge-myq2)

# Homebridge myQ<SUP STYLE="font-size: smaller; color:#5EB5E6;">2</SUP>

[![Downloads](https://badgen.net/npm/dt/homebridge-myq2)](https://www.npmjs.com/package/homebridge-myq2)
[![Version](https://badgen.net/npm/v/homebridge-myq2)](https://www.npmjs.com/package/homebridge-myq2)
[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)

## myQ garage door and other myQ-enabled device support for [Homebridge](https://homebridge.io).
</SPAN>

`homebridge-myq2` is a [Homebridge](https://homebridge.io) plugin that makes myQ-enabled devices available to [Apple's](https://www.apple.com) [HomeKit](https://www.apple.com/ios/home) smart home platform. myQ-enabled devices include many smart garage door openers made primarily by Liftmaster, Chamberlain, and Craftsman, but includes other brands as well. You can determine if your garage door or other device is myQ-enabled by checking the [myQ compatibility check tool](https://www.myq.com/myq-compatibility) on the myQ website.

There are two ways to control a myQ-compatible garage door opener through [HomeKit](https://www.apple.com/ios/home):

1. Liftmaster and Chamberlain make a hardware HomeKit bridge also called [Home Bridge](https://www.liftmaster.com/myq-home-bridge/p/G819LMB) (not to be confused with the open source [Homebridge project](https://homebridge.io)).
Unfortunately, some of us have encountered significant issues with the hardware bridge in a real world setting, where it either stops working or hangs for extended periods of time. That said, other users have encountered no issues and this hardware solution works well.

2. A plugin for [Homebridge](https://homebridge.io) like this one that emulates the capabilities of a myQ [HomeKit](https://www.apple.com/ios/home) bridge device.

Either solution will provide you with robust HomeKit integration, and you'll soon be automating your myQ smart garage with the richness of Apple's HomeKit ecosystem!

## Why use this plugin for myQ support in HomeKit?
In a nutshell, the aim of this plugin for things to *just work* with minimal required configuration by you, the end user. The goal is to provide as close to a streamlined experience as you would expect from a first-party or native HomeKit solution. For the adventurous, those additional granular options are, of course, available to support more esoteric use cases or other unique needs.

What does *just work* mean in practice? It means that this plugin will discover all your myQ devices and poll at regular, reasonable intervals for changes in state of a garage door opener or other myQ devices and inform HomeKit of those changes. By default. Without additional configuration beyond the login information required for myQ services.

### Features
- ***Easy* configuration - all you need is your myQ username and password to get started.** The defaults work for the vast majority of users. When you want more, there are [advanced options](#advanced-config) you can play with, if you choose.

- **Automatic detection and configuration of all garage door and gate openers.** By default - all of your supported myQ devices are made available in HomeKit.

- **[Obstruction detection](#obstruction-status) on supported myQ garage door and gate openers.** When a garage door or gate is obstructed, and the myQ API provides that information, you'll see an alert raised in the Home app.

- **[Battery status detection](#battery-status) on supported myQ door position sensor devices.** If you have a myQ supported door position sensor, you'll see an alert raised in the Home app to inform you when the battery is running low.

- **The ability to [selectively hide and show](#feature-options) specific gateways (useful when you have multiple homes) or openers.** For those who only want to show particular devices in HomeKit, or particular homes, a flexible and intuitive way to configure device availability at a granular level is available.

### <A NAME="myq-contribute"></A>How you can contribute and make this plugin even better
The myQ API is undocumented and implementing a plugin like this one is the result of many hours of reverse engineering, trial and error, and community support. This work stands on the shoulders of other myQ API projects out there and this project attempts to contribute back to that community base of knowledge to further improve myQ support for everyone.

I would love to support more types of myQ devices. I'm actively interested in working with people who have devices beyond garage doors in the myQ ecosystem such as myQ-enabled:

- Lights and lamps
- Motion sensors
- Cameras
- Locks

If you have these devices and would like to contribute, please open an [issue](https://github.com/hjdhjd/homebridge-myq2/issues), label it as a enhancement, and let's figure out how to make this plugin even better! Bonus points if you like puzzles and lots of debugging output. :smile:

## Installation
If you are new to Homebridge, please first read the [Homebridge](https://homebridge.io) [documentation](https://github.com/homebridge/homebridge/wiki) and installation instructions before proceeding.

If you have installed the [Homebridge Config UI](https://github.com/oznu/homebridge-config-ui-x), you can intall this plugin by going to the `Plugins` tab and searching for `homebridge-myq2` and installing it.

If you prefer to install `homebridge-myq2` from the command line, you can do so by executing:

```sh
sudo npm install -g homebridge-myq2
```

### Changelog
Changelog starting with v2.0 is available [here](https://github.com/hjdhjd/homebridge-myq2/blob/master/CHANGELOG.md).

### Things to be aware of
- **Beginning with v2.0, this plugin requires Homebridge v1.0 on greater to work. For some, this may be a breaking change if you are running on older versions of Homebridge.**

- Also beginning with v2.0, the `platform` configuration block for this plugin in your Homebridge `config.json` has been renamed (and note that the name is case sensitive as well). See the [plugin configuration section below](#plugin-configuration) for details. **This is a breaking change for those upgrading from v1.x and you will need to update your `config.json` to reflect the updates**.

- <A NAME="myq-errors"></A>The myQ API gets regularly updated and unfortunately this results in regularly breaking this and other myQ-related plugins. I've refactored this plugin in part to make it easier to maintain with future API changes that may come. Unfortunately, it's an ongoing challenge since API changes can be sudden and unpredictable.

- **As a result of the above you *will* see errors similar to this on an occasional basis in the Homebridge logs:**

    ```
    myQ API: Unable to update device status from myQ servers. Acquiring a new security token and retrying later.
    ```
  These messages can be safely ignored. myQ API errors *will* inevtiably happen. The myQ server-side infrastructure from Liftmaster / Chamberlain is not completely reliable and occasionally errors out due to server maintenance, network issues, or other infrastructure hiccups that occur on the myQ end of things. This plugin has no control over this, unfortunately, and all we can do is handle those errors gracefully, which is what I've attempted to do. The logging is informative and not a cause for significant concern unless it is constant and ongoing, which would be indicative of the larger API issues referenced above. When one of these errors is detected, we log back into the myQ infrastructure, obtain new API security credentials, and attempt refresh our status in the next scheduled update, which by is roughly [every 12 seconds by default](#advanced-config).

- <A NAME="obstruction-status"></A>Obstruction detection in myQ is more nuanced than one might think at first glance. When myQ detects an obstruction, that obstruction is only visible in the API for a *very* small amount of time, typically no more than a few seconds. This presents a user experience problem - if you remain completely faithful to the myQ API and only show the user the obstruction for the very short amount of time that it actually occurs, the user might never notice it because the alert is not visible for more than a few seconds. Instead, the design decision I've chosen to make is to ensure that any detected obstruction is alerted in HomeKit for 30 seconds from the last time myQ detected that obstruction. This ensures that the user has a reasonable chance of noticing there was an obstruction at some point in the very recent past, without having to have the user stare at the Home app constantly to happen to catch an ephemeral state.

- <A NAME="battery-status"></A>If your myQ device has support for battery status, `homebridge-myq2` will automatically detect and add support for it in HomeKit. However, you **will** see a warning message in the [Homebridge](https://homebridge.io) logs along the lines of:
    ```
    HAP Warning: Characteristic 00000079-0000-1000-8000-0026BB765291 not in required or optional characteristics for service 00000041-0000-1000-8000-0026BB765291. Adding anyway.
    ```
  This can be safely ignored. It's an error message indicating that, in HomeKit, the garage door opener accessory service doesn't normally support battery status. HomeKit will still report it correctly, and alert you accordingly.

## Plugin Configuration
If you choose to configure this plugin directly instead of using the [Homebridge Configuration web UI](https://github.com/oznu/homebridge-config-ui-x), you'll need to add the platform to your `config.json` in your home directory inside `.homebridge`.

```js
"platforms": [{
    "platform": "myQ",
    "email": "email@email.com",
    "password": "password"
}]
```

For most people, I recommend using [Homebridge Configuration web UI](https://github.com/oznu/homebridge-config-ui-x) to configure this plugin rather than doing so directly. It's easier to use for most users, especially newer users, and less prone to typos, leading to other problems.

### Feature Options
Feature options allow you to enable or disable certain features in this plugin.

The `options` setting is an array of strings used to customize feature options. Available options:

* <CODE>Hide.<I>serialnumber</I></CODE> - hide the opener or gateway identified by `serialnumber` from HomeKit.
* <CODE>Show.<I>serialnumber</I></CODE> - show the opener or gateway identified by `serialnumber` from HomeKit.

The plugin will log all devices it encounters and knows about, and you can use that to guide what you'd like to hide or show.

Before using this feature, you should understand how gateways and openers work in myQ. Gateways are the devices in your home that actually communicate your status to myQ. Openers are attached to gateways. A typical home will have a single gateway and one, or more, openers. If you choose to hide a gateway, you will also hide all the openers associated with that gateway.

If you've hidden a gateway, and all it's openers with it, you can selectively enable a single opener associated with that gateway by explicitly setting a `Show.` feature option. This should give you a lot of richness in how you enable or disable devices for HomeKit use.

The priority given to these options works in this order, from highest to lowest priority where settings that are higher in priority can override lower ones:

* Show any opener we've explicitly said to show.
* Show any gateway we've explicitly said to show.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This project uses [semantic versioning](https://semver.org/).

## 2.1.8 (2020-07-25)
  * Additional web UI refinements.

## 2.1.6, v2.1.7 (2020-07-25)
  * Fix: remove spurious and noisy log entry when polling the myQ API.

## 2.1.5 (2020-07-25)
  * **Wanted: more myQ device types. If you have a myQ light or other non-door myQ accessory, I'd love to [hear from you](https://github.com/hjdhjd/homebridge-myq2#myq-contribute) and see if we can add support for it in `homebridge-myq2`.**
  * Enhancement: alert on obstructions for a longer window of time (30 seconds), to give users a better chance of noticing them. See [here](https://github.com/hjdhjd/homebridge-myq2#obstruction-status) if you'd like to read more about it.
  * Enhancement: refine stopped state to better inform users when the state occurs.<BR>*Note: the default iOS Home app doesn't seem to correctly show a garage door in a stopped state, however other HomeKit apps such as Eve Home will correctly show this state when it occurs.*
  * Enhancement: increase our update resolution when we detect any state change, not just when we initiate one.
  * Enhancement: tweaked default myQ API refresh intervals to provide more frequent state updates, within reason.
  * Change: [some options have been renamed](https://github.com/hjdhjd/homebridge-myq2#advanced-config), and one new option has been added - the ability to override the builtin myQ application identifier. **Use with extreme caution and only if you know what you are doing.**
  * Update debug logging approach across the codebase.
  * Miscellaneous housekeeping edits and code maintenance.
  * Updated documentation and logo.

## 2.1.4 (2020-07-20)
  * Fix: address a race condition in updating device status.
  * Logging is more standardized and refined across the plugin.
  * Deprecated and removed `openDuration` and `closeDuration`. The values have been unused internally for some time, and they don't materially impact polling resolution.
  * Prepare the plumbing for additional myQ devices.

## 2.1.3 (2020-07-17)

  * Fix: refresh security tokens more often to address potential myQ API issues.
  * Get a status update from myQ immediately on startup.
  * Remove reachability support since it's now deprecated in HomeKit and homebridge.
  * Refine logging to clarify messages and streamline in places.
  * Minor updates to the code base.

## 2.1.2 (2020-07-12)

  * Fix: repair npm install script.

## 2.1.1 (2020-07-12)

  * Enhancement: deduce the type of device and brand based on serial number.
  * Enhancement: inform users when we choose not to add a device to HomeKit because we don't support it yet.
  * Fix: don't attempt to open or close the door if we're already in that state.
  * Fix: acquire a new myQ API security token regularly (thanks @dxdc for helping track this one down).
  * Fix: address a potential race condition when we check for battery information availability (on supported models).

## 2.1.0 (2020-07-12)

  * Feature: include battery status information for devices that support it.
  * Code cleanup.

## 2.0.13 (2020-07-11)
## 2.0.12 (2020-07-11)

  * Fix: look at the `device_family` attribute to determine whether it's a garage opener or not, rather than the `device_type` attribute.

## 2.0.11

  * New feature: feature options. This replaces the previous gateways and openers settings and should be a bit more intuitive to use.

## 2.0.10

  * Improved state handling for opening and closing conditions, including dealing with edge cases.
  * Preserve door state information across homebridge instances, so we remember where we left off.
  * myQ API cleanup.

## 2.0.1 - 2.0.9 (2020-07-04)

  * API fixes to ensure compatibility.
  * Re-include UI-based configuration.
  * Re-include README and CHANGELOG.
  * Broaden our filtering for garage door openers (who knew there were so many types?!) :smile:

  Thanks to [shamoon](https://github.com/shamoon) and others for debugging and contributing to the API fixes and troubleshooting.


## 2.0.0 (2020-07-03)

  ### Breaking Changes

  * Plugin requires homebridge >= 1.0.0.
  * This plugin has been refactored to typescript.
  * Update to myQ API v5.1.
  * Configuration changes:
	* Platform name has changed to `myQ`. **This will break existing configurations, so ensure you regenerate or update your `config.json` accordingly**.
	* The settings `gateways` and `openers` still exist but currently do nothing. This will be fixed in a future release.
	* Battery status is no longer provided as it doesn't seem to exist in the most recent myQ API. **If you were using this feature, please open an issue and the author can work with you to determine if the API exposes this functionality and make it available in this plugin**.


index.js
var fetch = require("node-fetch");
var Accessory, Service, Characteristic, UUIDGen;

module.exports = function (homebridge) {
  Accessory = homebridge.platformAccessory;
  Service = homebridge.hap.Service;
  Characteristic = homebridge.hap.Characteristic;
  UUIDGen = homebridge.hap.uuid;

  homebridge.registerPlatform("homebridge-myq2", "MyQ2", MyQ2Platform, true);
}

// This seems to be the "id" of the official LiftMaster iOS app
var APP_ID = "JVM/G9Nwih5BwKgNCjLxiFUQxQijAebyyg8QUHr7JOrP+tuPb8iHfRHKwTmDzHOu";
var UA_ID = "myQ/19859 CFNetwork/1107.1 Darwin/19.0.0";

// Headers needed for validation
var HEADERS = {
    "Content-Type": "application/json",
    "User-Agent": UA_ID,
    "BrandID": "2",
    "ApiVersion": "5.1",
    "Culture": "en",
    "MyQApplicationID": APP_ID
};

function MyQ2Platform(log, config, api) {
  this.log = log;
  this.config = config || {"platform": "MyQ2"};
  this.debug = this.config.debug === true;
  this.email = this.config.email;
  this.password = this.config.password;
  this.gateways = Array.isArray(this.config.gateways) ? this.config.gateways : [];
  this.openers = Array.isArray(this.config.openers) ? this.config.openers : [];
  this.openDuration = parseInt(this.config.openDuration, 10) || 15;
  this.closeDuration = parseInt(this.config.closeDuration, 10) || 25;
  this.longPoll = parseInt(this.config.longPoll, 10) || 15;
  this.shortPoll = parseInt(this.config.shortPoll, 10) || 5;
  this.shortPollDuration = parseInt(this.config.shortPollDuration, 10) || 600;
  this.maxCount = this.shortPollDuration / this.shortPoll;
  this.count = this.maxCount;
  this.validData = false;
  this.accountID;

  // Gateways convenience
  if(this.config.gateways) this.gateways.push(this.config.gateways);
  if(this.config.openers) this.gateways.push(this.config.openers);
  if(this.config.openers && Array.isArray(this.config.openers)) {
    this.gateways = this.gateways.concat(this.config.openers);
  }

  this.accessories = [];

  if(api) {
    this.api = api;
    this.api.on('didFinishLaunching', this.didFinishLaunching.bind(this));
  }

  // Definition Mapping
  this.doorState = ["open.", "closed.", "opening.", "closing.", "stopped."];
  this.batteryState = ["normal.", "low."];
}

// Method to restore accessories from cache
MyQ2Platform.prototype.configureAccessory = function (accessory) {
  this.setService(accessory);
  this.accessories[accessory.context.deviceID] = accessory;
}

// Method to setup accesories from config.json
MyQ2Platform.prototype.didFinishLaunching = function () {
  if(this.email && this.password) {
    // Add or update accessory in HomeKit
    this.addAccessory();

    // Start polling
    this.statePolling(0);
  } else {
    this.log("Error: MyQ username and password is required.");
    for (var deviceID in this.accessories) {
      var accessory = this.accessories[deviceID];
      this.removeAccessory(accessory);
    }
  }
}

// Method to add or update HomeKit accessories
MyQ2Platform.prototype.addAccessory = function() {
  var self = this;

  this.login(function (error){
    if(!error) {
      for (var deviceID in self.accessories) {
        var accessory = self.accessories[deviceID];
        if(!accessory.reachable) {
          // Remove extra accessories in cache
          self.removeAccessory(accessory);
        } else {
          // Update inital state
          self.log("Initializing platform accessory '" + accessory.context.name + " (ID: " + deviceID + ")'...");
          self.updateDoorStates(accessory);
        }
      }
    }
  });
}

// Method to remove accessories from HomeKit
MyQ2Platform.prototype.removeAccessory = function(accessory) {
  if(accessory) {
    var deviceID = accessory.context.deviceID;
    this.log(accessory.context.name + " is removed from HomeBridge.");
    this.api.unregisterPlatformAccessories("homebridge-myq2", "MyQ2", [accessory]);
    delete this.accessories[deviceID];
  }
}

// Method to setup listeners for different events
MyQ2Platform.prototype.setService = function(accessory) {
  accessory.getService(Service.GarageDoorOpener)
    .getCharacteristic(Characteristic.CurrentDoorState)
    .on('get', this.getCurrentState.bind(this, accessory.context));

  accessory.getService(Service.GarageDoorOpener)
    .getCharacteristic(Characteristic.TargetDoorState)
    .on('get', this.getTargetState.bind(this, accessory.context))
    .on('set', this.setTargetState.bind(this, accessory.context));

  accessory.getService(Service.GarageDoorOpener)
    .getCharacteristic(Characteristic.StatusLowBattery)
    .on('get', this.getStatusLowBattery.bind(this, accessory.context))

  accessory.on('identify', this.identify.bind(this, accessory));
}

// Method to setup HomeKit accessory information
MyQ2Platform.prototype.setAccessoryInfo = function (accessory, model, serial) {
  accessory.getService(Service.AccessoryInformation)
    .setCharacteristic(Characteristic.Manufacturer, this.manufacturer)
    .setCharacteristic(Characteristic.Model, model)
    .setCharacteristic(Characteristic.SerialNumber, serial);
}

// Method to update door state in HomeKit
MyQ2Platform.prototype.updateDoorStates = function (accessory) {
  accessory.getService(Service.GarageDoorOpener)
    .setCharacteristic(Characteristic.CurrentDoorState, accessory.context.currentState);

  accessory.getService(Service.GarageDoorOpener)
    .getCharacteristic(Characteristic.TargetDoorState)
    .getValue();

  accessory.getService(Service.GarageDoorOpener)
    .setCharacteristic(Characteristic.StatusLowBattery, accessory.context.batteryStatus);
}

// Method to retrieve door state from the server
MyQ2Platform.prototype.updateState = function (callback) {
  if(this.validData) {
    // Refresh data directly from sever if current data is valid
    this.getDevice(callback);
  } else {
    // Re-login if current data is not valid
    this.login(callback);
  }
}

// Method for state periodic update
MyQ2Platform.prototype.statePolling = function (delay) {
  var self = this;
  var refresh = this.longPoll + delay;

  // Clear polling
  clearTimeout(this.tout);

  // Determine polling interval
  if(this.count  < this.maxCount) {
    this.count++;
    refresh = this.shortPoll + delay;
  }

  // Setup periodic update with polling interval
  this.tout = setTimeout(function () {
    self.updateState(function (error) {
      if(!error) {
        // Update states for all HomeKit accessories
        for (var deviceID in self.accessories) {
          var accessory = self.accessories[deviceID];
          self.updateDoorStates(accessory);
        }
      } else {
        // Re-login after short polling interval if error occurs
        self.count = self.maxCount - 1;
      }

      // Setup next polling
      self.statePolling(0);
    });
  }, refresh * 1000);
}

// Login to MyQ server
MyQ2Platform.prototype.login = function (callback) {
  var self = this;

  // Body stream for validation
  var body = {
    UserName: this.email,
    Password: this.password
  };

  // login to Liftmaster
  fetch("https://api.myqdevice.com/api/v5/Login", {
    method: "POST",
    headers: HEADERS,
    body: JSON.stringify(body)
  }).then(function (res) {
    if (self.debug) self.log.debug('login response status: ' + res.status);
    return res.json();
  }).then(function (data) {
    if (self.debug) self.log.debug('login data:', data);
    if (data.SecurityToken) {
      self.securityToken = data.SecurityToken;
      self.manufacturer = "Chamberlain";
      // Adding security token to headers
      var getHeaders = JSON.parse(JSON.stringify(HEADERS));
      getHeaders.SecurityToken = self.securityToken;

      // set account on first login
      if (!self.accountID) {
        fetch("https://api.myqdevice.com/api/v5/My", {
          method: "GET",
          headers: getHeaders
        }).then(function (res) {
          if (self.debug) self.log.debug('My response status: ' + res.status);
          return res.json();
        }).then(function (data) {
          if (self.debug) self.log.debug('My response data:', data);
          if (data.Account) {
            self.accountID = data.Account.href.substring(data.Account.href.lastIndexOf('/') + 1);
            self.getDevice(callback);
          } else {
            self.log.error(`${data.message}: ${data.description}`);
            callback(data.message);
            return;
          }
        });
      } else { // already have account
        self.getDevice(callback);
      }
    } else {
      self.log.error('Unable to login to MyQ:', data.message);
      callback(data.message);
    }
  }).catch(error => {
      self.log.error('Unable to login to MyQ, received error:', error);
      callback(error);
  });
}

// Find your garage door ID
MyQ2Platform.prototype.getDevice = function (callback) {
  var self = this;

  // Reset validData hint until we retrived data from the server
  this.validData = false;

  // Querystring params
  var query = {
    filterOn: "true"
  };

  // Adding security token to headers
  var getHeaders = JSON.parse(JSON.stringify(HEADERS));
  getHeaders.SecurityToken = this.securityToken;
  // Request details of all your devices
  fetch(`https://api.myqdevice.com/api/v5.1/Accounts/${self.accountID}/Devices`, {
    method: "GET",
    headers: getHeaders,
    query: query
  }).then(function (res) {
    if (self.debug) self.log.debug('getDevice response status: ' + res.status);
    return res.json();
  }).then(function (data) {
    if (self.debug) self.log.debug('getDevice response data:', data);
    if (data.count > 0) {
      var devices = data.items;

      // Handle MyQ fetch errors gracefully. This is especially helpful when MyQ
      // API changes happen.
      if (devices === undefined) {
        var parseErr = "Error: Couldn't fetch device details:\r\n" + JSON.stringify(data, null, 2);
        self.log(parseErr);
        callback(parseErr);
        return;
      }

      // Look through the array of devices for all the gateways
      var allowedGateways = [];
      var gatewaysKeyed = [];

      for (var i = 0; i < devices.length; i++) {
        var device = devices[i];
        var deviceType = device.device_type;
        var deviceDesc = device.name;

        // Is this gateway one of the specified gateways in the config
        gatewaysKeyed[device.serial_number] = deviceDesc;
        if(self.gateways.indexOf(deviceDesc) > -1 || self.gateways.indexOf(device.serial_number) > -1) allowedGateways.push(device.serial_number);
      }

      // Look through the array of devices for all the openers
      for (var i = 0; i < devices.length; i++) {
        var device = devices[i];
        var deviceType = device.device_type;

        // Search for specific device type
        if(deviceType == 'virtualgaragedooropener' || deviceType == "garagedooropener" || deviceType === "gate") {
          var thisDeviceID = device.serial_number.toString();
          var thisModel = deviceType.toString();
          var thisDoorName = device.name;
          var thisDoorState = device.state.door_state;
          var thisDoorMonitor = device.state.monitor_only_mode == 'true';
          var thisDoorBatteryLow = device.state.dps_low_battery_mode == 'true';

          // Does this device fall under the specified gateways
          if(self.gateways.length > 0 && allowedGateways.indexOf(device.parent_device_id) == -1) {
            if(self.debug) {
              self.log('Skipping Device: "'+thisDoorName+'" - Device ID: '+thisDeviceID+' (Gateway: "'+gatewaysKeyed[device.parent_device_id]+"\"",'-', "Gateway ID:",device.parent_device_id+")");
            }

            continue;
          }

          // Does this device fail under the specified openers
          if(self.openers.length > 0 && self.openers.indexOf(device.serial_number) == -1) {
            if(self.debug) {
              self.log('Skipping Device: "'+thisDoorName+'" - Device ID: '+thisDeviceID+' (Gateway: "'+gatewaysKeyed[device.parent_device_id]+"\"",'-', "Gateway ID:",device.parent_device_id+")");
            }

            continue;
          }

          if (!thisDoorMonitor) {
            // Retrieve accessory from cache
            var accessory = self.accessories[thisDeviceID];

            // Initialization for new accessory
            if(!accessory) {

              // Setup accessory as GARAGE_DOOR_OPENER (4) category.
              var uuid = UUIDGen.generate(thisDeviceID);
              accessory = new Accessory("MyQ " + thisDoorName, uuid, 4);

              // Setup HomeKit security system service
              accessory.addService(Service.GarageDoorOpener, thisDoorName);

              // New accessory is always reachable
              accessory.reachable = true;

              // Setup HomeKit accessory information
              self.setAccessoryInfo(accessory, thisModel, thisDeviceID);

              // Setup listeners for different security system events
              self.setService(accessory);

              // Register new accessory in HomeKit
              self.api.registerPlatformAccessories("homebridge-myq2", "MyQ2", [accessory]);

              // Store accessory in cache
              self.accessories[thisDeviceID] = accessory;
            }

            if(self.debug) {
              if(device.parent_device_id) {
                self.log('Checking "'+thisDoorName+'" - Device ID: '+thisDeviceID+' (Gateway: "'+gatewaysKeyed[device.parent_device_id]+"\"",'-', "Gateway ID:",device.parent_device_id+")");
              } else {
                self.log('Checking: "'+thisDoorName+'"');
              }
            }

            // Accessory is reachable after it's found in the server
            accessory.updateReachability(true);

            // Store and initialize variables into context
            var cache = accessory.context;
            cache.name = thisDoorName;
            cache.deviceID = thisDeviceID;
            if (cache.currentState == undefined)
              cache.currentState = Characteristic.CurrentDoorState.CLOSED;
            if (cache.batteryStatus == undefined)
              cache.batteryStatus = Characteristic.StatusLowBattery.BATTERY_LEVEL_NORMAL;

            // Determine the current door state
            // TODO: v5 API may not support all these states
            var newState;
            if (thisDoorState == 'open') {
              newState = Characteristic.CurrentDoorState.OPEN;
            } else if (thisDoorState == 'closed') {
              newState = Characteristic.CurrentDoorState.CLOSED;
            } else if (thisDoorState == 'stopped') {
              newState = Characteristic.CurrentDoorState.STOPPED;
            } else if (thisDoorState == 'opening') {
              newState = Characteristic.CurrentDoorState.OPENING;
            } else if (thisDoorState == 'closing') {
              newState = Characteristic.CurrentDoorState.CLOSING;
            } else {
              // Not sure about this...
              accessory.updateReachability(false);
            }

            // Determine current battery state
            var newBattery;
            if (thisDoorBatteryLow) {
              newBattery = Characteristic.StatusLowBattery.BATTERY_LEVEL_LOW;
            } else {
              newBattery = Characteristic.StatusLowBattery.BATTERY_LEVEL_NORMAL;
            }

            // Detect for state changes
            // Just roll them together for now, battery state changes
            // should be very infrequent.
            if(newState !== cache.currentState ||
               newBattery != cache.batteryStatus) {
              self.count = 0;
              cache.currentState = newState;
              cache.batteryStatus = newBattery;
              self.log(cache.name + " is " + self.doorState[cache.currentState] + " Battery " + self.batteryState[cache.batteryStatus]);
            }

            // Set validData hint after we found an opener
            self.validData = true;

            // Ensure the accessories cache is updated to avoid registryy issues
            self.api.updatePlatformAccessories([accessory]);
          }
        }
      }

      // Did we have valid data?
      if(self.validData) {
        // Set short polling interval when state changes
        self.statePolling(0);

        callback();
      } else {
        var parseErr = "Error: Couldn't find a MyQ door device."
        self.log.error(parseErr);
        callback(parseErr);
      }
    } else {
      self.log.error("Error getting MyQ devices:", data.message, data.description);
      callback(data.message);
    }
  }).catch(error => {
      self.log.error('Error polling MyQ servers:', error);
      callback(error);
  });
}

// Send opener target state to the server
MyQ2Platform.prototype.setState = function (thisOpener, state, callback) {
  var self = this;
  var thisAccessory = this.accessories[thisOpener.deviceID];
  var myqState = state == Characteristic.CurrentDoorState.CLOSED ? 'close' : 'open';
  var updateDelay = state == Characteristic.CurrentDoorState.CLOSED ? this.closeDuration : this.openDuration;

  // Adding security token to headers
  var putHeaders = JSON.parse(JSON.stringify(HEADERS));
  putHeaders.SecurityToken = this.securityToken;

  // PUT request body
  var body = {
    action_type: myqState
  };

  // Send the state request to Liftmaster
  fetch(`https://api.myqdevice.com/api/v5.1/accounts/${self.accountID}/devices/${thisOpener.deviceID}/actions`, {
    method: "PUT",
    headers: putHeaders,
    body: JSON.stringify(body)
  }).then(function (res) {
    if (res.ok) { // v5 API just response status 204
      self.log(thisOpener.name + " is set to " + self.doorState[state]);

      // Set short polling interval
      self.count = 0;
      self.statePolling(updateDelay - self.shortPoll);

      callback();
    } else {
      self.log.error("Error setting " + thisOpener.name + " state:", res);
      callback(res);
    }
  }).catch(error => {
      self.log.error('Error setting the target: ' + error);
      callback(error);
  });
}

// Method to set target door state
MyQ2Platform.prototype.setTargetState = function (thisOpener, state, callback) {
  var self = this;

  // Always re-login for setting the state
  this.login(function (loginError) {
    if(!loginError) {
      self.setState(thisOpener, state, callback);
    } else {
      callback(loginError);
    }
  });
}

// Method to get target door state
MyQ2Platform.prototype.getTargetState = function (thisOpener, callback) {
  // Get target state directly from cache
  callback(null, thisOpener.currentState % 2);
}

// Method to get current door state
MyQ2Platform.prototype.getCurrentState = function (thisOpener, callback) {
  var self = this;

  // Retrieve latest state from server
  this.updateState(function (error) {
    if(!error) {
      self.log(thisOpener.name + " is " + self.doorState[thisOpener.currentState] + " Battery " + self.batteryState[thisOpener.batteryStatus]);
      callback(null, thisOpener.currentState);
    } else {
      callback(error);
    }
  });
}

// Method to get battery status
MyQ2Platform.prototype.getStatusLowBattery = function (thisOpener, callback) {
  // Get battery status directly from cache
  callback(null, thisOpener.batteryStatus);
}

// Method to handle identify request
MyQ2Platform.prototype.identify = function (thisOpener, paired, callback) {
  this.log(thisOpener.name + " identify requested!");
  callback();
}
